rsl/rsl.go

```diff
@@ -0,0 +1,198 @@
+/*
+AddRSLEntry adds an RSL entry to the log for the ref and hash passed in.
+TODO: Do we just use commit signing for signing RSL entries?
+*/
+func AddRSLEntry(refName string, commitID plumbing.Hash, sign bool) error {
+	repo, err := common.GetRepositoryHandler()
+	if err != nil {
+		return err
+	}
+	message := fmt.Sprintf("%s: %s", refName, commitID.String())
+
+	return commit(repo, plumbing.ZeroHash, RSLRef, message, sign)
+}
+
+/*
+GetLatestRSLEntry returns the latest entry available locally in the RSL.
+TODO: There is no information yet about the signature for the entry.
+*/
+func GetLatestRSLEntry() (string, plumbing.Hash, error) {
+	repo, err := common.GetRepositoryHandler()
+	if err != nil {
+		return "", plumbing.ZeroHash, err
+	}
+
+	ref, err := repo.Reference(plumbing.ReferenceName(RSLRef), true)
+	if err != nil {
+		return "", plumbing.ZeroHash, err
+	}
+
+	commitObj, err := repo.CommitObject(ref.Hash())
+	if err != nil {
+		return "", plumbing.ZeroHash, err
+	}
+
+	messageSplit := strings.Split(commitObj.Message, ":")
+
+	return strings.Trim(messageSplit[0], " "),
+		plumbing.NewHash(strings.Trim(messageSplit[1], " ")), nil
+}
+
+func commit(repo *git.Repository, treeHash plumbing.Hash, targetRef string,
+	message string, sign bool) error {
+	gitConfig, err := repo.ConfigScoped(config.GlobalScope)
+	if err != nil {
+		return err
+	}
+
+	curRef, err := repo.Reference(plumbing.ReferenceName(targetRef), true)
+	if err != nil {
+		return err
+	}
+
+	author := object.Signature{
+		Name:  gitConfig.User.Name,
+		Email: gitConfig.User.Email,
+		When:  time.Now(),
+	}
+
+	commit := object.Commit{
+		Author:    author,
+		Committer: author,
+		TreeHash:  treeHash,
+		Message:   message,
+	}
+	if curRef.Hash() != plumbing.ZeroHash {
+		commit.ParentHashes = []plumbing.Hash{curRef.Hash()}
+	}
+
+	if sign {
+		signingMethod, keyInfo, program, err := common.GetSigningInfo(repo)
+		if err != nil {
+			return err
+		}
+		signature, err := signCommit(repo, &commit, signingMethod,
+			keyInfo, program)
+		if err != nil {
+			return err
+		}
+		commit.PGPSignature = signature
+	}
+
+	obj := repo.Storer.NewEncodedObject()
+	err = commit.Encode(obj)
+	if err != nil {
+		return err
+	}
+	commitHash, err := repo.Storer.SetEncodedObject(obj)
+	if err != nil {
+		return err
+	}
+
+	newRef := plumbing.NewHashReference(plumbing.ReferenceName(targetRef),
+		commitHash)
+	return repo.Storer.CheckAndSetReference(newRef, curRef)
+}
+
+func signCommit(repo *git.Repository, commit *object.Commit, signingMethod common.SigningMethod,
+	keyInfo string, program string) (string, error) {
+	commitEncoded := repo.Storer.NewEncodedObject()
+	if err := commit.EncodeWithoutSignature(commitEncoded); err != nil {
+		return "", err
+	}
+	r, err := commitEncoded.Reader()
+	if err != nil {
+		return "", err
+	}
+	commitContents, err := io.ReadAll(r)
+	if err != nil {
+		return "", err
+	}
+
+	switch signingMethod {
+	case common.SigningMethodGPG:
+		cmd := exec.Command(program,
```

Comment:
@wlynch @znewman01 any thoughts on how we invoke the default program for commit signing here? I may have overcomplicated it a bit so curious to hear your thoughts. Also, this doesn't work, it just pauses here. I suspect it's either an issue with how the input is passed on to gpg (I can see that the process is invoked) or it's because the password prompt isn't passed on to me. Thoughts?

Reply:
you might have to mess with 'pinentry'? the GPG CLI is a bit of a mess and unfortunately not super easy to call as a library either

Reply:
It was an oversight on my part with how I was passing in the contents to be signed using stdin. :smile: